namesys/repubns.go

```diff
@@ -0,0 +1,62 @@
+package namesys
+
+import (
+	"context"
+	"time"
+
+	path "github.com/ipfs/go-ipfs/path"
+
+	ds "gx/ipfs/QmPpegoMqhAEqjncrzArm7KVWAkCm78rqL2DPuNjhPrshg/go-datastore"
+	routing "gx/ipfs/QmTiWLZ6Fo5j4KcTVutZJ5KWRRJrbxzmxA4td8NfEdrPh7/go-libp2p-routing"
+	peer "gx/ipfs/QmZoWKhxUmZ2seW4BzX6fJkNR8hh9PsGModr7q171yq2SS/go-libp2p-peer"
+	ci "gx/ipfs/QmaPbCnUMBohSGo3KnxEa2bHqyJVVeEEcwtqJAYxerieBo/go-libp2p-crypto"
+	dshelp "gx/ipfs/QmdQTPWduSeyveSxeCAte33M592isSW5Z979g81aJphrgn/go-ipfs-ds-help"
+)
+
+// repubns extends mpns to store published records in the datastore so
+// that they will be picked up by the republisher
+type repubns struct {
+	mpns
+	dstore ds.Datastore
+}
+
+// NewRepublishingNameSystem will construct an extension of the naming system that
+// also republishes nodes
+func NewRepublishingNameSystem(r routing.ValueStore, ds ds.Datastore, cachesize int) NameSystem {
+	ns := NewNameSystem(r, ds, cachesize)
+	return &repubns{*ns.(*mpns), ds}
+}
+
+// Publish publishes a name / value
+func (ns *repubns) Publish(ctx context.Context, name ci.PrivKey, value path.Path) error {
+	return ns.PublishWithEOL(ctx, name, value, time.Now().Add(DefaultRecordTTL))
+}
+
+// Publish publishes a name / value with an EOL
+func (ns *repubns) PublishWithEOL(ctx context.Context, name ci.PrivKey, value path.Path, eol time.Time) error {
+	err := ns.mpns.PublishWithEOL(ctx, name, value, eol)
+	if err != nil {
+		return err
+	}
+
+	id, err := peer.IDFromPrivateKey(name)
+	if err != nil {
+		return err
+	}
+
+	_, ipnskey := IpnsKeysForID(id)
+	rec, err := ns.dstore.Get(dshelp.NewKeyFromBinary([]byte(ipnskey)))
```

Comment:
Making assumptions about how the `NameSystem` stores this information is really unfortunate. Having to store everything twice is also really unfortunate.

My original plan was to simply record "we've published record `/ipns/QmId => /ipfs/QmThing`" at time X then call `Publish(name, value)` every X + RepubTime (recording the new time).

However, this brings up a *different* issue: this isn't limited to republishing. We actually need to remember the last sequence number for namesys to work properly. Therefore, while this PR fixes the immediate issue, `NewNameSystem` still returns a broken namesys.

Reply:
I agree it is unfortunate to make assumptions about how `NameSystem` stores the information. There are instances all over the namesys code of reaching directly into the datastore to get a record. It makes me wonder if the routing interface should have a `GetLocalValue(key)` method that doesn't go out to the DHT or do any validation of records.

I'm not sure I understood your last comment. Are you saying that namesys republishing is broken, because we need to get the last sequence number for a record, but we only look locally instead of going out to the network?

Reply:
Yes, but, more than that, the underlying namesys implementation itself is broken (as it will forget the last sequence number).

Reply:
Sorry I'm still not sure I follow... why would it forget the last sequence number? Is the datastore not persistent?

Reply:
The problem is that the DHT is now deleting expired entries, regardless of whether or not they're *our* expired entry. That means that when namesys tries to get them directly from the datastore, it may not find them.

Although.... arguably, the sequence number doesn't matter for expired records. That is, if the last known record *has* expired, we can safely start over from sequence number 0. Thoughts @whyrusleeping?

In that case, the republisher would simply store a mapping of `RepublisherKey(ipnskey) => last_publish` and namesys wouldn't interact with datastores *at all*. Instead, namesys would call `GetValue` on the `routing.ValueStore` to determine if there exists an unexpired record and use the sequence number from that. That solves all the problems with services digging into eachother's datastores.

Reply:
What you're suggesting makes sense, and as I understand it that's basically what Publish() does: it [gets the last sequence number](https://github.com/ipfs/go-ipfs/blob/master/namesys/publisher.go#L61) by [looking in the local datastore](https://github.com/ipfs/go-ipfs/blob/master/namesys/publisher.go#L73), and if it doesn't find the record there it [calls GetValue()](https://github.com/ipfs/go-ipfs/blob/master/namesys/publisher.go#L96) on the DHT. If there's no record in the DHT it uses sequence number zero. Maybe republish should do the same thing?

Are you suggesting we remove the check for the record in the local datastore from `Publish()` to avoid having to look directly into the datastore? I assume it's there for performance reasons.

Note also that when there is a GET RPC on the DHT, it will [send the highest sequence number record](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/routing.go#L119) to peers who have an IPNS record with a lower sequence number, so over time the highest sequence number records should gradually replace lower sequence number records throughout the DHT.

What does `last_publish` represent exactly? I'm just wondering about what happens in the case when a node goes down for a couple of days and then comes back up and wants to republish its (now expired) records. Or are we not concerned about that use case?

Reply:
@Stebalien one issue I see with maintaining the sequence number in `Namesys` rather than within the DHT is that (if I understand correctly) keys can be shared between peers, so if another peer is publishing ipns records with the same key as my peer then my peer will get out of sync on the sequence numbers. One solution is to always call `GetValue()` but that has performance implications, which is presumably why `Publish()` looks directly in the datastore.

Reply:
> Are you suggesting we remove the check for the record in the local datastore from Publish() to avoid having to look directly into the datastore? I assume it's there for performance reasons.

Yes.

It's probably there so we don't have to go and ask the network. However, for correctness, we should probably be doing that anyways. There's no guarantee (or even reason to believe) that we're the only one publishing the IPNS record. Honestly, it shouldn't be *that* expensive given that we're going to be doing a DHT put to the *exact* same nodes immediately after calling `GetValue`.

> Maybe republish should do the same thing?

I was thinking that republish would just call the normal `Publish` function and let that handle everything. Note: to avoid bumping the sequence number unnecessarily, publish can avoid incrementing it when the value doesn't change. 

> Note also that when there is a GET RPC on the DHT, it will send the highest sequence number record to peers who have an IPNS record with a lower sequence number, so over time the highest sequence number records should gradually replace lower sequence number records throughout the DHT.

Unless I'm mistaken, this should only apply to non-expired IPNS records. When we're dealing with non-expired records, asking the DHT for the record should work just fine.

> What does last_publish represent exactly? I'm just wondering about what happens in the case when a node goes down for a couple of days and then comes back up and wants to republish its (now expired) records. Or are we not concerned about that use case?

We are. Sorry, my comment there was wrong (part of an alternative proposal that I rejected). The republisher does need to record the last *value* published. That is, `RepublisherKey(ipnskey) => /ipfs/QmLastValue`.

Note: I'd *like* to, additionally, store the last time we published every record (or, alternatively, the last time we did a republish). That's where that `last_publish` came from. At the moment, we *always* republish 5 minutes after starting, regardless of when we last republished as we never know how long we've been offline. However, we can do this separately (and probably just store a global "last republish" value).

Reply:
I'm not sure if the republisher should store the last value published. For example in this scenario:
- Peer A publishes an IPNS record to its own key
- Peer A shares its key with Peer B
- Peer B publishes an IPNS record to Peer A's key (with a higher sequence number)
The value in Peer A's datastore will be updated by Peer B's DHT PUT, however Peer A's republisher's value for key A is now out of date.

Reply:
Yeah... However, at that point, I'm starting to wonder if we shouldn't just merge the republisher with namesys and avoid having separate systems entirely. That would probably simplify this anyways.

Reply:
Yes I think that might make things simpler.

I believe there's actually a bug with the current implementation that we should fix as well. It seems like the republisher will periodically [republish the values for _all_ the keys](https://github.com/ipfs/go-ipfs/blob/master/namesys/republisher/repub.go#L103) in its possession, and it [only checks its local datastore](https://github.com/ipfs/go-ipfs/blob/master/namesys/republisher/repub.go#L127) before republish. That could cause a problem in the following scenario:
- Peer A publishes an IPNS record to Peer B's key with sequence number 5. It is retained in Peer A's local DHT datastore.
- Peer B publishes an IPNS record to its own key with sequence number 6.
- Peer A will not receive Peer B's DHT PUT, because Peer A is not one of the closest peers for key B
- When Peer A republishes it checks its local datastore and finds the record with sequence number 5, which it then republishes to key B
- Peer B _will_ receive Peer A's DHT PUT, overwriting the record with sequence number 6 in its local datastore (and in all the closest peers to key B)

I think the solution here again is for the republisher to always go out to the DHT with `GetValue()` before publishing

Reply:
> Peer B will receive Peer A's DHT PUT, overwriting the new record (sequence number 6) with the old record (sequence number 5) in its local datastore (and in all the closest peers to key B)

Peer B should ignore this put as the sequence number is lower (unless peer B's value has expired).

But I agree, we should always call GetValue first.

Reply:
It looks like the handler will [check that the record is valid](https://github.com/libp2p/go-libp2p-kad-dht/blob/master/handlers.go#L175) (ie, signature is ok and record is not expired) but it will not compare the sequence number of the new record with its record before overwriting its local copy

Reply:
That's... horribly wrong. https://github.com/libp2p/go-libp2p-kad-dht/issues/121